servicetalk-http-router-jersey/src/testFixtures/java/io/servicetalk/http/router/jersey/CancellationTest.java

```diff
@@ -258,7 +259,9 @@ public void onError(final Throwable t) {
 
         cancelledLatch.await();
 
-        assertThat(errorRef.get(), is(nullValue()));
+        if (errorRef.get() != null) {
```

Comment:
The test failure is racy but you can see that an exception is regularly generated (see below). Some times the cancel (which closes the EntityInputStream) happens before the `DefaultJerseyStreamingRouter` calls `applicationHandler.handle(containerRequest);`, and when jersey tries to `readEntity` it observes the stream is closed. Getting this error shouldn't be a problem, but my bigger concern is are we interacting with the `EntityInputStream` in a thread safe way according to the APIs (e.g. `DefaultContainerResponseWriter#cancelResponse()` initiates the close and is protected by an atomic, but jersey is reading from the stream too which maybe on a different thread IIUC?).

```
java.lang.IllegalStateException: Entity input stream has already been closed.
	at org.glassfish.jersey.message.internal.EntityInputStream.ensureNotClosed(EntityInputStream.java:205)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:865)
	at org.glassfish.jersey.server.ContainerRequest.readEntity(ContainerRequest.java:274)
	at org.glassfish.jersey.server.internal.inject.EntityParamValueParamProvider$EntityValueSupplier.apply(EntityParamValueParamProvider.java:73)
	at org.glassfish.jersey.server.internal.inject.EntityParamValueParamProvider$EntityValueSupplier.apply(EntityParamValueParamProvider.java:56)
	at org.glassfish.jersey.server.spi.internal.ParamValueFactoryWithSource.apply(ParamValueFactoryWithSource.java:50)
	at org.glassfish.jersey.server.spi.internal.ParameterValueHelper.getParameterValues(ParameterValueHelper.java:68)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$AbstractMethodParamInvoker.getParamValues(JavaResourceMethodDispatcherProvider.java:109)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter.handle0(DefaultJerseyStreamingHttpRouter.java:251)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter.access$100(DefaultJerseyStreamingHttpRouter.java:70)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter$3.handleSubscribe(DefaultJerseyStreamingHttpRouter.java:183)
	at io.servicetalk.concurrent.api.Single.handleSubscribe(Single.java:1955)
	at io.servicetalk.concurrent.api.Single.lambda$subscribeWithContext$12(Single.java:1931)
	at io.servicetalk.concurrent.api.ContextPreservingConsumer.accept(ContextPreservingConsumer.java:40)
	at io.servicetalk.concurrent.api.NoopOffloader.offloadSubscribe(NoopOffloader.java:74)
	at io.servicetalk.concurrent.api.Single.subscribeWithContext(Single.java:1930)
	at io.servicetalk.concurrent.api.Single.subscribeAndReturnContext(Single.java:1885)
	at io.servicetalk.concurrent.api.Single.subscribeInternal(Single.java:1336)
	at io.servicetalk.concurrent.api.SingleFlatMapSingle$SubscriberImpl.onSuccess(SingleFlatMapSingle.java:81)
	at io.servicetalk.concurrent.api.ContextPreservingSingleSubscriber.onSuccess(ContextPreservingSingleSubscriber.java:75)
	at io.servicetalk.concurrent.api.AbstractSubmitSingle.lambda$handleSubscribe$0(AbstractSubmitSingle.java:49)
	at io.servicetalk.concurrent.api.ContextPreservingRunnable.run(ContextPreservingRunnable.java:43)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
Thanks for the stack-trace that helps a lot.

The problem here is that the `RequestContainer` itself isn't thread safe, and our cancellable calls `request.close()` from the caller thread (whatever that is), whereas the `readEntity` runs from the offloaded thread.
Managed to consistently reproduce on my dev env with a latch awaiting in `cancel` and released by `InboundMessageContext.readEntity`.